### PR TITLE
Add support for `pip>=21.3`

### LINKED ISSUE #1503: pip-sync fails to run on pip >= 21.3
It seems that pip 21.3 has removed `get_installed_distributions` from `_internal.utils.misc`. `pip-sync` is importing this, and that is causing a catastrophic error on startup.

#### Environment Versions

1. Linux
1. Python: 3.8.6
1. pip version: 21.3
1. pip-tools version: 6.3.1

#### Steps to replicate

1. Create a virtual environment and activate it
2. `pip install -U pip`
3. `pip install pip-tools`
4. `pip-sync -h`

#### Expected result

Help

#### Actual result

```
Traceback (most recent call last):
  File "/home/smmorton/playground/demo/bin/pip-sync", line 5, in <module>
    from piptools.scripts.sync import cli
  File "/home/smmorton/playground/demo/lib/python3.8/site-packages/piptools/scripts/sync.py", line 12, in <module>
    from pip._internal.utils.misc import get_installed_distributions
ImportError: cannot import name 'get_installed_distributions' from 'pip._internal.utils.misc' (/home/smmorton/playground/demo/lib/python3.8/site-packages/pip/_internal/utils/misc.py)
```

### PR DESCRIPTION
**Pull request by atugushev**
Changes:
- fix removed in https://github.com/pypa/pip/commit/d051a00fc57037104fca85ad8ebf2cdbd1e32d24 `pip._internal.utils.misc.get_installed_distributions` used in `piptools.scripts.sync`

TODO:

- [x] merge #1500 
- [x] rebase/merge-master

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

Fixes #1503

### CODE REVIEW

**Comment by atugushev** [comment]
@jazzband/pip-tools now this is all set up and ready for review.

**Comment by atugushev** [comment]
Thanks @ssbarnea 🙏🏻

File: piptools/utils.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import json
 import os
 import shlex
+import typing
```
Comment (AndydeCleyre) [inline piptools/utils.py:6]: I know I'm late, but importing `typing` directly (and later using `typing.Union` rather than just `Union`) seems unnecessary.

**Comment by rrauenza** [comment]
thanks for the super quick fix, everyone!  Saw this breaking across my cicd's and I was beginning to pin pip version...

**Pull request merged**